runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -802,33 +788,40 @@ protected void unmapSecondaryIndexes(K key, V value) {
      */
     @SuppressWarnings("unchecked")
     protected void mapSecondaryIndexes(K key, V value) {
+        if (value == null) {
+            log.trace("Attempting to build an index with a null value, skipping.");
+            return;
+        }
+
         try {
-            if (value != null) {
-                // Map entry into secondary indexes
-                for (Index<K, V, ? extends Comparable> index : indexSpec) {
-                    String indexName = index.getName().get();
-                    Map<Comparable, Map<K, V>> secondaryIndex = secondaryIndexes.get(indexName);
-                    if (index.isMonoIndex()) {
-                        Comparable indexKey = index.getIndexFunction().apply(key, value);
-                        Map<K, V> slot = secondaryIndex.computeIfAbsent(indexKey, k -> new HashMap<>());
-                        slot.put(key, value);
-                    } else {
-                        for (Comparable<?> indexKey
-                                : index.getMultiValueIndexFunction().apply(key, value)) {
-                            Map<K, V> slot = secondaryIndex
-                                    .computeIfAbsent(indexKey, k -> new HashMap<>());
-                            slot.put(key, value);
-                        }
-                    }
+            // Map entry into secondary indexes
+            for (Index<K, V, ? extends Comparable> index : indexSpec) {
+                String indexName = index.getName().get();
+                Map<Comparable, Map<K, V>> secondaryIndex = secondaryIndexes.get(indexName);
+                for (Comparable indexKey : index.getMultiValueIndexFunction().apply(key, value)) {
+                    Map<K, V> slot = secondaryIndex.computeIfAbsent(indexKey, k -> new HashMap<>());
+                    slot.put(key, value);
                 }
             }
         } catch (Exception e) {
-            indexSpec.clear();
-            secondaryIndexes.clear();
-            indexGenerationFailed = true;
-            log.error("mapSecondaryIndexes: Exception mapping {}, {},"
-                    + " UNMAPPING ALL INDEXES, indexing is disabled", key, value, e);
+            log.error("Received an exception while computing the index. " +
+                    "This is most likely an issue with the client's indexing function. {}", e);
+            // The index might be corrupt, and the only way to ensure safety is to disable
+            // indexing for this table.
+            clearIndex();
+            // In case of both a transactional and non-transactional operation, the client
+            // is going to receive UnrecoverableCorfuError along with the appropriate cause.
+            throw e;
         }
     }
 
+    /**
+     *  Disable all seconday indecies for this table. Only used during error-recovery.
+     */
+    protected void clearIndex() {
+        indexSpec.clear();
+        secondaryIndexes.clear();
+        indexGenerationFailed = true;
```

Comment:
should `indexGenerationFailed` be volatile or AtomicBoolean?

Reply:
Also what about `indexSpec` and `secondaryIndexes` ConcurrentSet and ConcurrentMap respectively?

Reply:
indexSpec and secondaryIndexes are actually fine, since their mutation is tied to CorfuTable construction.

I don't think that indexGenerationFailed is used at all. I'll remove it.

Reply:
ok